### PR TITLE
JavaScript Carousel | Jenny Tran

### PR DESCRIPTION
**Pull request by jennydtran**
Nov-05-2020 09-29-36

### CODE REVIEW

File: javascript-carousel/main.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  nextButton.classList.add('current-button');
+}
+
+// click right arrow moves image reel forward
+function clickRightChevron(event) {
+  var rightArrow = document.querySelector('.fa-chevron-right');
+
+  if (event.target !== rightArrow) {
+    return;
+  }
```
Comment (thebearingedge) [inline javascript-carousel/main.js:41]: You have the right idea here using a guard to stop the function. But if possible, avoid querying the DOM during event handler functions. You can combine these lines by using [`Element.matches()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches).

```js
if (!event.target.matches('.fa-chevron-right')) {
  return;
}
```

File: javascript-carousel/main.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  var leftArrow = document.querySelector('.fa-chevron-left');
+  var currentImage = document.querySelector('.current-pokemon');
+  var currentButton = document.querySelector('.current-button');
+
+  var previousImage = currentImage.previousElementSibling;
+  var previousButton = currentButton.previousElementSibling;
+
+  if (event.target !== leftArrow) {
+    return;
+  }
```
Comment (thebearingedge) [inline javascript-carousel/main.js:60]: Try to get guards out of the way immediately at the top of the function so you don't needlessly query the DOM. Try `Element.matches()`as mentioned before.

File: javascript-carousel/main.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  clearTimeout(timerID);
+  timerID = setTimeout(imageChangeWithInterval, 1300);
+}
+document.addEventListener('click', clickLeftChevron);
+
+// click on circle buttons to change to any image
+function clickButtonForImages(event) {
+  if (event.target.tagName !== 'BUTTON') {
+    return;
+  }
```
Comment (thebearingedge) [inline javascript-carousel/main.js:84]: 👍

**Review by thebearingedge — CHANGES_REQUESTED** [review]
Good job so far. We're going to take some steps to simplify your logic, but start with these couple of changes first.

**Review by thebearingedge — CHANGES_REQUESTED** [review]
Ok, great. Now we need to fix a strategic issue in your code. As of now, your code relies on exactly what order the pictures are in and what their `id` attributes are. That's problematic because if you wanted to show another picture or change their order, you'd have to update your HTML _and_ JavaScript.

Try this:
1. Define a function named `displayImage(index)`. It should take an `index` number and loop through _all_ images and _all_ dots and "show" the right one according to `index`. To timeout involved no event listeners. If you comment out line 11, you should be able to test this function by hand in your browser console.
1. Declare a variable named `activeIndex`.
1. Define a function named `getNextIndex()` that returns you the "next" index or `0` depending on what `activeIndex` is. Again, this should be testable in the console. It doesn't update the DOM or set any timeouts or anything, just returns a number.
1. Define a function named `getPreviousIndex()` that returns you the "previous" index or the last one. Also should be testable in the console.

Once you have these pieces, send me a message and we'll tie it together. If you want, you should be able to use these things to get your auto-play and left/right icons to function properly.

File: javascript-carousel/main.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+var activeIndex = 0;
+
+function displayImage(index) {
+  for (var i = 0; i < imageList.length; i++) {
+    if (i !== index) {
+      imageList[i].classList.add('hidden');
+      buttonsList[i].classList.remove('current-button');
+    }
+    imageList[index].classList.remove('hidden');
+    buttonsList[index].classList.add('current-button');
```
Comment (thebearingedge) [inline javascript-carousel/main.js:22]: There should be an `else` in here for this function to work properly.

File: javascript-carousel/main.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    nextButton = document.querySelector('#firstbutton');
+  }
+
+  currentImage.classList.remove('current-pokemon');
+  currentImage.classList.add('hidden');
+  nextImage.classList.remove('hidden');
+  nextImage.classList.add('current-pokemon');
+
+  currentButton.classList.remove('current-button');
+  nextButton.classList.add('current-button');
```
Comment (thebearingedge) [inline javascript-carousel/main.js:60]: Ok, in here you can get the next index and display the proper image with your functions.

File: javascript-carousel/main.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    previousButton = document.querySelector('#lastbutton');
+  }
+
+  currentImage.classList.remove('current-pokemon');
+  currentImage.classList.add('hidden');
+  previousImage.classList.remove('hidden');
+  previousImage.classList.add('current-pokemon');
+
+  currentButton.classList.remove('current-button');
+  previousButton.classList.add('current-button');
```
Comment (thebearingedge) [inline javascript-carousel/main.js:98]: You can use your functions here too, to get the previous index and display the correct image.

File: javascript-carousel/main.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  var currentImage = document.querySelector('.current-pokemon');
+
+  for (var i = 0; i < buttonsList.length; i++) {
+    if (event.target === buttonsList[i]) {
+      currentImage.classList.remove('current-pokemon');
+      currentImage.classList.add('hidden');
+      imageList[i].classList.remove('hidden');
+      imageList[i].classList.add('current-pokemon');
+    }
+  }
```
Comment (thebearingedge) [inline javascript-carousel/main.js:128]: Here you just need to figure out the index of the event target in the button list and call your `displayImage(i)` function.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: javascript-carousel/main.js (hunk 7 above)
Comment (jennydtran) [reply javascript-carousel/main.js]: Tim, thanks for all the guidance!

**Review by thebearingedge — APPROVED** [review]

**Pull request merged**